MoltenVK/MoltenVK/GPUObjects/MVKSwapchain.mm

```diff
@@ -408,15 +415,38 @@ static inline CIE1931XY VkXYColorEXTToCIE1931XY(VkXYColorEXT xy) {
 VkResult MVKSwapchain::getRefreshCycleDuration(VkRefreshCycleDurationGOOGLE *pRefreshCycleDuration) {
 	if (_device->getConfigurationResult() != VK_SUCCESS) { return _device->getConfigurationResult(); }
 
+#if MVK_IOS_OR_TVOS || MVK_MACCAT
 	NSInteger framesPerSecond = 60;
-#if MVK_IOS_OR_TVOS
 	UIScreen* screen = [UIScreen mainScreen];
 	if ([screen respondsToSelector: @selector(maximumFramesPerSecond)]) {
 		framesPerSecond = screen.maximumFramesPerSecond;
 	}
 #endif
-#if MVK_MACOS
-	// TODO: hook this up for macOS, probably need to use CGDisplayModeGetRefeshRate
+
+#if MVK_MACOS && !MVK_MACCAT
+	// Find the screen for the window whose content view is backed by the swapchains CAMetalLayer.
+	// Default to the mainScreen if no such window can be found.
+	NSScreen* screen = [NSScreen mainScreen];
+	CALayer* layer = _mtlLayer;
+	while (layer.superlayer) {
+		layer = layer.superlayer;
+	}
+
+	for (NSWindow* window in [[NSApplication sharedApplication] windows]) {
+		NSView *view = [window contentView];
+		if (view && ([view layer] == layer)) { // Check against layer and not _mtlLayer.
+			screen = [window screen];
+		}
+	}
+
+	CGDirectDisplayID displayId = [[[screen deviceDescription] objectForKey:@"NSScreenNumber"] unsignedIntValue];
+	CGDisplayModeRef mode = CGDisplayCopyDisplayMode(displayId);
+	double framesPerSecond = CGDisplayModeGetRefreshRate(mode);
+	CGDisplayModeRelease(mode);
+
+	// Builtin panels, e.g., on MacBook, report a zero refresh rate.
+	if (framesPerSecond == 0)
+		framesPerSecond = 60.0;
```

Comment:
A note for the future: I believe some newer builtin displays are capable of higher refresh rates. (I *know* some iPad displays have higher refresh rates, but those don't run macOS.) At some point, we may want to find a way to query what that is. There must be some way, or otherwise `CVDisplayLink` wouldn't know how much to limit frame rates to.

Reply:
> we may want to find a way to query what that is

Would [`UIScreen:: maximumFramesPerSecond`](https://developer.apple.com/documentation/uikit/uiscreen/2806814-maximumframespersecond?language=objc) take care of it? 

@spnda If it's that simple, can we add it here before pulling this in?

Reply:
UIScreen::maximumFramesPerSecond is already there in the iOS/tvOS etc. code-path. But that doesn't exist on macOS.

Reply:
I can confirm that my M1 MacBook Pro has the "ProMotion" display which in theory is an "adaptive" 120Hz screen. It's pretty unclear to me what that actually means from an API standpoint though, i.e. whether it's something like FreeSync or whether we have any API visibility into opting into or out of this feature from an application standpoint.

Reply:
Hmmm...looks like the equivalent in macOS 12+ might be [`NSScreen::minimumrefreshinterval `](https://developer.apple.com/documentation/appkit/nsscreen/3824747-minimumrefreshinterval?language=objc).

If we include this, we'll have to wrap the call in a `#if MVK_XCODE_13` macro.

Reply:
If it helps, looks like GLFW used to [solve for this with for this with CVDisplayLink](https://github.com/glfw/glfw/blob/b4a8eb9b197caebcec1304cd3d55cd56e3869376/src/cocoa_monitor.m#L239) before replacing it with some wild IOKit heroics

Reply:
> Hmmm...I was close! 😉 
> 
> 
> 
> Looks like the equivalent in macOS 12+ is [`NSScreen:: maximumFramesPerSecond.`](https://developer.apple.com/documentation/appkit/nsscreen/3824747-minimumrefreshinterval?language=objc) It was introduced to support Active-Sync displays.
> 
> 
> 
> If we include this, we'll have to wrap the call in a `#if MVK_XCODE_13` macro.

Yeah, I'll add this then.